### PR TITLE
docs: add documentation on Check algorithm

### PR DESCRIPTION
**Pull request by jon-whit**
## Description
This adds some documentation to our `docs` that describes how the internal Check resolution algorithm works inside of the `internal/graph` package.

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: docs/check/overview.md [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+* **Subproblem** - an FGA query that is contingent on one or more evaluations of nested relationship evaluations (including itself).
+
+* **Dispatch** - refers to the process of evaluating a nested subproblem in order to resolve some parent subproblem.
+
+## Direct Rewrites
+Direct rewrites establish the most basic form of FGA relationship rewrite definitions. These direct rewrites allow developers to express the most basic of authorization model semantics, including direct permissions, indirect or computed permissions, and/or hierarchical permissions.
+
+> ℹ️ The term "permission" is used above, but from here forward we will formally refer to permissions simply as relationships. A users/subject has a permission if they have the relationship, and a relationship can be realized through one or more relationship rewrites.
+
+### Direct Relationships
```
Comment (miparnisari) [inline docs/check/overview.md:43]:
```suggestion
### Model with Direct Relationships. Example Check
```

here and elsewhere. I skipped through all these thinking that they were going to explain what these terms meant, only to realize that these actually have examples of Check resolutions :)

File: docs/check/overview.md [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+- [] todo: fill me out
+
+#### Resolution Breadth
+- [] todo: fill me out
+
+#### Bounding Concurrency at the Storage Layer
+- [] todo: fill me out
+
+### CheckResolver
+The [graph#CheckResolver](https://github.com/openfga/openfga/blob/7bdf3398b47a96995cb0877f25b065a7f6f5a8e1/internal/graph/interface.go#L9) interface establishes a contract that all implementations of a Check query resolver must implement. The interface allows for layered dispatch composition as well as remote dispatch semantics (though this is not implemented yet).
```
Comment (miparnisari) [inline docs/check/overview.md:536]: we should add godocs to that interface 😆

**Comment by poovamraj** [comment]
Just starting to work on this doc. I'd rename this from overview.md to README.md as Github directly opens files named README when we enter the folder as shown here - https://github.com/openfga/openfga/tree/concurrency-controls-in-check-docs/docs/check

**Review by poovamraj — COMMENTED** [review]
Just starting to work on this doc. I'd rename this from overview.md to README.md as Github directly opens files named README when we enter the folder as shown here - https://github.com/openfga/openfga/tree/concurrency-controls-in-check-docs/docs/check

File: docs/check/overview.md [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+
+## Definitions
+Familiarity with basic [OpenFGA Concepts](https://openfga.dev/docs/concepts) is expected, but the following are some terms used throughout this document that are more specific to the content herein:
+
+* **Relation Rewrite (Rewrite)** - a rule that expresses an FGA relationship as a composition of other relationships. For example, you can "rewrite" a viewer relationship as editor or express a hierarchy between two objects. These rules can also be composed using set operations including union, intersection, and exclusion.
```
Comment (miparnisari) [inline docs/check/overview.md:28]: what do you think about including a drawing showing all the possible rewrites? (add union, intersection and exclusion)

image

File: docs/check/overview.md [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+
+* **Rewrite Operand** - for rewrites involving set operations (e.g. union, intersection, exclusion) a rewrite operand refers to the relation rewrites that are evaluated as part of the set expression. For example, the rewrite operands of the definition `a or b` are `a` and `b`.
+
+* **Expansion** - refers to the <u>process</u> of iteratively evaluating one or more subproblems in order to determine some FGA query outcome. 
+
+* **Subproblem** - an FGA query that is contingent on one or more evaluations of nested relationship evaluations (including itself).
+
+* **Dispatch** - refers to the process of evaluating a nested subproblem in order to resolve some parent subproblem.
```
[the remark this answers is not in the document]
Comment (jon-whit) [reply docs/check/overview.md]: The term "expansion" in the scope of this Check doc is not to be confused with the Expand API. I've added some more text that hopefully clarifies the distinction when it comes to the semantics of how we use the term "expand" vs "dispatch" when we talk about FGA queries. Let me know what you think with the updates 👍

File: docs/check/overview.md (hunk 1 above)
Comment (jon-whit) [reply docs/check/overview.md:78]: We do explain the term "Direct Relationships" below. Specifically, we say
> If a relation can be directly assigned or given/shared with a specific user/subject, then we refer to it as a "direct relationship". Direct relationships are any of those relationships that, for a particular authorization model, can be written to an [FGA Store](https://openfga.dev/docs/concepts#what-is-a-store) using the [Write API](https://openfga.dev/api/service#/Relationship%20Tuples/Write).

The content of the doc is organized by rewrite rule categories. There are 2 categories - direct rewrite rules and set rewrite rules. The set rewrite rules compose other set rewrite rules and/or direct rewrite rules. Self (e.g. direct relationships), computed userset, and TTU encompass the direct rewrite rules. Union, intersection, and exclusion encompass the set rewrite rules.

The purpose of this section was to a) outline the direct rewrite rules and what they do, specifically in this subsection for direct relationships and b) provide examples of them. I feel like the content as is adequately portrays the intent.

File: docs/check/overview.md [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+## Definitions
+Familiarity with basic [OpenFGA Concepts](https://openfga.dev/docs/concepts) is expected, but the following are some terms used throughout this document that are more specific to the content herein:
+
+* **Relation Rewrite (Rewrite)** - a rule that expresses an FGA relationship as a composition of other relationships. For example, you can "rewrite" a viewer relationship as editor or express a hierarchy between two objects. These rules can also be composed using set operations including union, intersection, and exclusion.
+
+* **Rewrite Operand** - for rewrites involving set operations (e.g. union, intersection, exclusion) a rewrite operand refers to the relation rewrites that are evaluated as part of the set expression. For example, the rewrite operands of the definition `a or b` are `a` and `b`.
+
+* **Expansion** - refers to the <u>process</u> of iteratively evaluating one or more subproblems in order to determine some FGA query outcome. 
+
+* **Subproblem** - an FGA query that is contingent on one or more evaluations of nested relationship evaluations (including itself).
```
[the remark this answers is not in the document]
Comment (jon-whit) [reply docs/check/overview.md]: I've reworded a bit, but kept the intent 👍

File: docs/check/overview.md (hunk 3 above)
Comment (jon-whit) [reply docs/check/overview.md]: I translated this to a diagram (including set rewrites) and have now included it 👍

File: docs/check/overview.md [hunk 6]
```diff
@@ -0,0 +1,583 @@
+# Check
+The [Check API](https://openfga.dev/api/service#/Relationship%20Queries/Check) looks up if a particular user/subject has specific relationship with a given object. 
+
+It is a forward expansion algorithm that starts at a given `object#relation` and iteratively and recursively expands relationships until a particular user/subject is found or until all paths of the expansion have been visited and no such paths were found. It can be viewed as a tree traversal or a directed graph traversal.
```
[the remark this answers is not in the document]
Comment (jon-whit) [reply docs/check/overview.md]: Added a simple diagram with a brief description. Hopefully that gives a quick idea of the general structure which can be useful to understanding the remainder of the content herein. 👍

File: docs/check/README.md [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+
+* [CachedCheckResolver](https://github.com/openfga/openfga/blob/7bdf3398b47a96995cb0877f25b065a7f6f5a8e1/internal/graph/cached_resolver.go#L67) - this implementation provides a local cache of subproblems that have already been evaluated in the recent past. We avoid evaluating the same subproblem we've recently seen and we re-use the cached result from the previous evaluation of that subproblem.
+
+* [LocalChecker](https://github.com/openfga/openfga/blob/7bdf3398b47a96995cb0877f25b065a7f6f5a8e1/internal/graph/check.go#L125) - the LocalChecker implements the core algorithm. The LocalChecker follows relationship rewrite rules, does database lookups, and executes the expansion and dispatch of subproblems. A LocalChecker also has a delegate that it dispatches subproblems to, which allows the LocalChecker to dispatch subproblems through one of the CheckResolvers mentioned above and avoids duplicating subproblem evaluation and/or re-evaluation of previously computed subproblems.
+
+These implementations are very purposefully layered like so:
+```
+SingleflightCheckResolver <---------------------|
```
Comment (miparnisari) [inline docs/check/README.md:590]: please remove all mentions of singleflight until we actually bring it in

File: docs/check/fga_model_highlighted.svg
Comment (miparnisari) [inline docs/check/fga_model_highlighted.svg:1]: love this drawing!

**Review by miparnisari — APPROVED** [review]

**Pull request merged**